src/command.rs

```diff
@@ -1,11 +1,130 @@
-use crate::entity::Entity;
-use crate::world::World;
+use crate::{
+    entity::Entity,
+    filter::{ChunksetFilterData, Filter},
+    storage::{Component, Tag},
+    world::{IntoComponentSource, TagLayout, TagSet, World},
+};
+use std::{marker::PhantomData, sync::Arc};
 
-pub enum EntityCommand {
-    Delete(Entity),
-    Create(Entity),
+trait WorldWritable {
+    fn write(self, world: &mut World);
 }
 
-pub struct EntityCommandBuffer {
+struct InsertCommand<T, C> {
+    tags: T,
+    components: C,
+}
+impl<T, C> WorldWritable for InsertCommand<T, C>
+where
+    T: TagSet + TagLayout + for<'a> Filter<ChunksetFilterData<'a>>,
+    C: IntoComponentSource,
+{
+    fn write(self, world: &mut World) { world.insert(self.tags, self.components); }
+}
+
+struct DeleteEntityCommand(Entity);
+impl WorldWritable for DeleteEntityCommand {
+    fn write(self, world: &mut World) { world.delete(self.0); }
+}
+
+struct AddTagCommand<T>(Entity, T);
+impl<T> WorldWritable for AddTagCommand<T>
+where
+    T: Tag,
+{
+    fn write(self, world: &mut World) { world.add_tag(self.0, self.1) }
+}
+
+struct RemoveTagCommand<T>(Entity, PhantomData<T>);
+impl<T> WorldWritable for RemoveTagCommand<T>
+where
+    T: Tag,
+{
+    fn write(self, world: &mut World) { world.remove_tag::<T>(self.0) }
+}
+
+struct AddComponentCommand<C>(Entity, C);
+impl<C> WorldWritable for AddComponentCommand<C>
+where
+    C: Component,
+{
+    fn write(self, world: &mut World) { world.add_component::<C>(self.0, self.1) }
+}
+
+struct RemoveComponentCommand<C>(Entity, PhantomData<C>);
+impl<C> WorldWritable for RemoveComponentCommand<C>
+where
+    C: Component,
+{
+    fn write(self, world: &mut World) { world.remove_component::<C>(self.0) }
+}
+
+enum EntityCommand {
+    WriteWorld(Arc<dyn WorldWritable>),
+}
+
+#[derive(Default)]
+pub struct CommandBuffer {
     commands: Vec<EntityCommand>,
 }
+impl CommandBuffer {
+    pub fn insert<T, C>(mut self, tags: T, components: C) -> Self
+    where
+        T: 'static + TagSet + TagLayout + for<'a> Filter<ChunksetFilterData<'a>>,
+        C: 'static + IntoComponentSource,
+    {
+        self.commands
+            .push(EntityCommand::WriteWorld(Arc::new(InsertCommand {
+                tags,
+                components,
+            })));
+
+        self
+    }
```

Comment:
Ideally, we would pre-allocate entity IDs and return them as `&[Entity]` in this call. This would require that the command buffer were constructed with a `&World` reference (so it can fetch entiy ID blocks from the world), which would make it difficult for the system to construct the buffer itself.

I also don't want to be allocating on the heap for each call, but rather use some kind of block allocator internally and then re-use that memory in later frames. This does not have to be implemented for the initial release, but it would again make it difficult for the system to construct the buffer itself.